### PR TITLE
Upgrade to Google Billing 5

### PR DESCRIPTION
**Pull request by mathkend**
Update to Google Billing 5
** Removed SKU Concept
** Purchases now can return multiple products. Only handling one (This can be enabled at Play Store - In App Product - Setting)
** The workflow about Subscriptions is unchanged and untested. 

The changes are tested in a similar way but not the actual code. Only changed to update the project since i do not use it anymore, but there are a lot of requests at the Discord. 

Feel free to question or improve the Pull Request.

### CODE REVIEW

**Comment by kharindev** [comment]
how to add this to my project? Thank you.

**Review by MrStahlfelge — COMMENTED** [review]
Thanks, I made some comments.
We upgraded Gradle as well, can you resolve conflicts caused by this when you are at it?

File: gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java [hunk 1]
```diff
@@ -225,37 +228,49 @@ public void dispose() {
...
-        List<Purchase> purchases = purchasesResult.getPurchasesList();
-
-        if (responseCode == BillingClient.BillingResponseCode.OK && purchases != null) {
-            handlePurchase(purchases, true);
-        } else {
-            Gdx.app.error(TAG, "queryPurchases failed with responseCode " + responseCode);
-            observer.handleRestoreError(new GdxPayException("queryPurchases failed with " +
-                    "responseCode " + responseCode));
-        }
+        mBillingClient.queryPurchasesAsync(
```
[the remark this answers is not in the document]
Comment (mathkend) [reply gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java:258]: queryPurchasesAsync and queryPurchaseHistoryAsync Are the only options i found to handle this. What would be your suggestion?
queryPurchasesAsync: Returns purchases details for currently owned items bought within your app.
queryPurchaseHistoryAsync: Returns the most recent purchase made by the user for each product, even if that purchase is expired, canceled, or consumed.

https://developer.android.com/reference/com/android/billingclient/api/BillingClient#queryPurchasesAsync(com.android.billingclient.api.QueryPurchasesParams,%20com.android.billingclient.api.PurchasesResponseListener)

Neither does return a full list of all your purchases correctly. I use a combination of both to handle errors during purchases.

File: gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java [hunk 2]
```diff
@@ -106,76 +106,79 @@ private String getGlobalSkuTypeFromConfig() {
...
-    private Information convertSkuDetailsToInformation(SkuDetails skuDetails) {
-        String priceString = skuDetails.getPrice();
+    private Information convertSkuDetailsToInformation(ProductDetails productDetails) {
+        String priceString = productDetails.getOneTimePurchaseOfferDetails().getFormattedPrice();
         return Information.newBuilder()
-                .localName(skuDetails.getTitle())
-                .freeTrialPeriod(convertToFreeTrialPeriod(skuDetails.getFreeTrialPeriod()))
-                .localDescription(skuDetails.getDescription())
+                .localName(productDetails.getTitle())
+                //TODO Removed - Dont know if this is applied in some other form
```
Comment (keesvandieren) [inline gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java:175]: This is needed for subscriptions. See https://qonversion.io/blog/google-play-billing-library-5-0/ for details

File: gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java [hunk 3]
```diff
@@ -106,76 +106,79 @@ private String getGlobalSkuTypeFromConfig() {
...
+                //.freeTrialPeriod(convertToFreeTrialPeriod(productDetails.getFreeTrialPeriod()))
+                .localDescription(productDetails.getDescription())
                 .localPricing(priceString)
-                .priceCurrencyCode(skuDetails.getPriceCurrencyCode())
-                .priceInCents((int) (skuDetails.getPriceAmountMicros() / 10_000))
-                .priceAsDouble(skuDetails.getPriceAmountMicros() / 1_000_000.0)
+                .priceCurrencyCode(productDetails.getOneTimePurchaseOfferDetails().getPriceCurrencyCode())
+                .priceInCents((int) (productDetails.getOneTimePurchaseOfferDetails().getPriceAmountMicros() / 10_000))
+                .priceAsDouble(productDetails.getOneTimePurchaseOfferDetails().getPriceAmountMicros() / 1_000_000.0)
                 .build();
```
Comment (keesvandieren) [inline gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java:182]: getOneTimePurchaesOfferDetails() is incorrect for subscriptions; productDetails.getSubscriptionOfferDetails() should be used, to be defined how to do it further

**Review by keesvandieren — CHANGES_REQUESTED** [review]
The updates breaks subscriptions. I will try to find some time to continue to work on it.

I am using Free trial period with one standard price afterwards, however Googles model supports much more models, which probably are hard to implement in the existing Information API.

Purchasing new subscriptions also has different API: https://qonversion.io/blog/google-play-billing-library-5-0/

I will continue on this tomorrow (in 22 hours from now)

File: gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (MrStahlfelge) [reply gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java:258]: Great, looks good. Just wanted to make sure.
Ping to @alex-dorokhov  for him to note and check when he updates.

File: gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java (hunk 2 above)
Comment (mathkend) [reply gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java:175]: Keep in mind that other parts also force CONSUMABLE product type. You may query your SUBSCRIPTION products, offer it, restore, etc.

File: gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java [hunk 4]
```diff
@@ -288,15 +303,17 @@ private void handlePurchase(List<Purchase> purchases, boolean fromRestore) {
         for (Purchase purchase : purchases) {
             // ignore pending purchases, just return successful purchases to the client
             if (purchase.getPurchaseState() == Purchase.PurchaseState.PURCHASED) {
+                //TODO - Only handling one product per purchase. This can be changed at Play Store - In app Product Settings
```
[the remark this answers is not in the document]
Comment (mathkend) [reply gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java:306]: Done

File: gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java [hunk 5]
```diff
@@ -106,76 +106,79 @@ private String getGlobalSkuTypeFromConfig() {
...
-                return BillingClient.SkuType.SUBS;
+                return BillingClient.ProductType.SUBS;
         }
         throw new IllegalStateException("Unsupported OfferType: " + type);
     }
 
-    private Information convertSkuDetailsToInformation(SkuDetails skuDetails) {
-        String priceString = skuDetails.getPrice();
+    private Information convertSkuDetailsToInformation(ProductDetails productDetails) {
+        String priceString = productDetails.getOneTimePurchaseOfferDetails().getFormattedPrice();
```
[the remark this answers is not in the document]
Comment (mathkend) [reply gdx-pay-android-googlebilling/src/com/badlogic/gdx/pay/android/googlebilling/PurchaseManagerGoogleBilling.java:172]: Done

**Comment by mathkend** [comment]
> how to add this to my project? Thank you.

Maybe you should wait for this to have a proper review, necessary changes and to be merged.

**Comment by mathkend** [comment]
> Thanks, I made some comments. We upgraded Gradle as well, can you resolve conflicts caused by this when you are at it?

@MrStahlfelge This changes in gradle requires Java 11, are you sure about it? (gdx-pay-android-amazon)

**Comment by MrStahlfelge** [comment]
> > Thanks, I made some comments. We upgraded Gradle as well, can you resolve conflicts caused by this when you are at it?
> 
> @MrStahlfelge This changes in gradle requires Java 11, are you sure about it? (gdx-pay-android-amazon)

Building Android needs Java 11 nowadays, so it is good to go.

**Comment by keesvandieren** [comment]
I have worked a while to support subscriptions, created patch based on your branch; basic support is there, but free trial period not yet. I have to find out how to simulate this, as on my device free trial period was already used. All help there is welcome.

[Add_subscriptions,_but_without_free_trial_info.patch.zip](https://github.com/libgdx/gdx-pay/files/9539684/Add_subscriptions._but_without_free_trial_info.patch.zip)

**Comment by bitdream** [comment]
Is  upgrading Gradle really necessary? Java 11 is not compatiple with Java 8! This is a big update.

**Comment by shpowley** [comment]
> I have worked a while to support subscriptions, created patch based on your branch; basic support is there, but free trial period not yet. I have to find out how to simulate this, as on my device free trial period was already used. All help there is welcome.
> 
> [Add_subscriptions,_but_without_free_trial_info.patch.zip](https://github.com/libgdx/gdx-pay/files/9539684/Add_subscriptions._but_without_free_trial_info.patch.zip)

I tested with your patch code. Thank you.

Individual CONSUMABLE and ENTITLEMENT purchases worked fine.
- Repeated purchases of CONSUMABLE worked fine.
- Repeated purchases of ENTITLEMENT reported as already owned.

Initially, SUBSCRIPTION with multiple offers _failed_ with my particular setup. Maybe because my initial offer was not "backwards compatible" (see attached screenshot). As per the recommendation in https://developer.android.com/google/play/billing/migrate-gpblv5#launching-offer, I only modified `getBillingFlowParams()` for ProductType and this fixed the error. My initial offer was a 1-month, pre-paid subscription and could be purchased multiple times.

```
    protected BillingFlowParams.Builder getBillingFlowParams(ProductDetails productDetails) {
        List<BillingFlowParams.ProductDetailsParams> productDetailsParamsList;

        if (productDetails.getProductType().equals(ProductType.INAPP)) {
            productDetailsParamsList =
              Collections.singletonList(
                BillingFlowParams.ProductDetailsParams.newBuilder()
                  .setProductDetails(productDetails)
                  .build()
              );
        }
        else {
            String offerToken = productDetails
              .getSubscriptionOfferDetails()
              .get(0) // HOW TO SPECIFY AN ALTERNATE OFFER USING gdx-pay?
              .getOfferToken();

            productDetailsParamsList =
              Collections.singletonList(
                BillingFlowParams.ProductDetailsParams.newBuilder()
                  .setProductDetails(productDetails)
                  .setOfferToken(offerToken)
                  .build()
              );
        }

        return BillingFlowParams.newBuilder().setProductDetailsParamsList(productDetailsParamsList);
    }
```

As mentioned in previous comments, only the SUBSCRIPTION's first offer can be purchased. I didn't know how to test free-trial scenario.

Screenshot of my play console subscription hierarchy:
Screenshot 2022-09-13 062925

**Comment by MrStahlfelge** [comment]
Great, thanks for testing! 
@keesvandieren can you push your changes, I think we can merge the branch then to publish a snapshot for more people to test.

**Comment by keesvandieren** [comment]
[Add_support_for_subscriptions_with_free_trial.patch.zip](https://github.com/libgdx/gdx-pay/files/9591155/Add_support_for_subscriptions_with_free_trial.patch.zip)

Attached the patch for support of subscriptions with free trial. 

I cannot test free trial as it is already used with my account and it is a once per account setting. But the code is simple, the complex logic is re-used from previous implementation so no risk there.

Can @mathkend  please apply this patch to your pull request branch? After that we can go for release I feel

**Comment by shpowley** [comment]
I tested the latest patch and the free trial appears to work.
- Initial subscription purchased _and free trial automatically was given_
- After the 5-min _test_ free trial period expired, I refunded the purchase
- For the same subscription, a subsequent purchase did not allow the free trial


I did encounter a couple issues.

1) As mentioned previously, hard-coding the subscription offer index(?) for index 0 causes problems. For instance, Play console doesn't allow deleting an offer - only de-activating the offer. An active offer at a higher index is inaccessible. Only solution is to just create a new subscription with the correct offer details..

2) Code only handles INFINITE_RECURRING subscriptions. Something like a pre-paid monthly subscription doesn't get added to `informationMap` (ex. only paying for only a month of video streaming). Just removing the `phase.getRecurrenceMode()` check fixed this.

OLD:
```
private static boolean isPaidForSubscriptionPhase(ProductDetails.PricingPhase phase) {
    return phase.getRecurrenceMode() == ProductDetails.RecurrenceMode.INFINITE_RECURRING && phase.getPriceAmountMicros() > 0;
}
```

NEW:
```
private static boolean isPaidForSubscriptionPhase(ProductDetails.PricingPhase phase) {
    return phase.getPriceAmountMicros() > 0;
}
```


UNRELATED:
Unclear how to apply .patch to PR. `$ git am <patch_file>` failed and manually applied changes.

**Comment by keesvandieren** [comment]
@shpowley thanks for your feedback! 

Regarding this:


> As mentioned previously, hard-coding the subscription offer index(?) for index 0 causes problems. For instance, Play console doesn't allow deleting an offer - only de-activating the offer. An active offer at a higher index is inaccessible. Only solution is to just create a new subscription with the correct offer details..

Do you have a suggestion of how to choose another? Maybe instead of get(0) we should always get the last one? Now I am using this method:


```
    private static ProductDetails.SubscriptionOfferDetails getActiveSubscriptionOfferDetails(List<ProductDetails.SubscriptionOfferDetails> subscriptionOfferDetails) {
        return subscriptionOfferDetails.get(subscriptionOfferDetails.size() - 1);
    }
```


Applied this method change:

```
    private static boolean isPaidForSubscriptionPhase(ProductDetails.PricingPhase phase) {
        return phase.getPriceAmountMicros() > 0;
    }
```

> UNRELATED:
> Unclear how to apply .patch to PR. $ git am <patch_file> failed and manually applied changes.

I use IDEA to Create the Patch via the Menu Git > Patch. Maybe apply from there is easier to use?

New version available here: [Add_support_for_subscriptions_with_free_trial.patch.zip](https://github.com/libgdx/gdx-pay/files/9615206/Add_support_for_subscriptions_with_free_trial.patch.zip)

Tested it with my subscription, worked for me. Waiting for your feedback.

**Comment by shpowley** [comment]
@keesvandieren 
Applied and tested latest patch directly against PR using Android Studio + GIT plugin (thanks).

Worked as expected for the different purchase scenarios mentioned in previous posts, but unable to test the free-trial again. Creating a new test app with duplicate code + Play console setup would trigger a free-trial, but that seems a bit overkill.


Regarding the subscription offer index. Play console allows you to activate/de-activate any offer in the list - so assuming that the first or last index is the intended/active one isn't ideal. An overloaded purchase() method on the PurchaseManager interface would be preferred I think.


In addition to the existing:

  `public void purchase(String identifier) {..}`


Also add:

  `public void purchase(String identifier, int subscriptionOfferIndex) {..}`

  - store `subscriptionOfferIndex` as a private class variable
  - fallback to index 0, if `subscriptionOfferIndex` is out-of-bounds?


Unfortunately, PurchaseManager is also implemented elsewhere and this is Google-specific issue. I tried placing the overloaded purchase() directly on PurchaseManagerGoogleBilling, but that didn't work. Is altering PurchaseManager not viable?

  HuaweiPurchaseManager
  PurchaseManagerAndroidAmazon
  PurchaseManageriOSApple

  - for these classes, just redirect to the existing purchase() method

**Comment by MrStahlfelge** [comment]
The offer index should be set in a Google-specific configuration.

**Comment by keesvandieren** [comment]
@shpowley will only enabled offerIndices be returned?  I checked for enabledness booleans in the APIs, but couldn't find any pointers there.

Can we release a first version where we cannot yet set the index, and do that in subsequent release?

**Comment by shpowley** [comment]
@keesvandieren In debug, the entire offer list is always returned, every base plan + child offer(s), regardless of active status. Multiple active offers can exist on a subscription. I didn't see a way determine an offer's active status. Deleting subscription data in Play console isn't supported.

I don't actually use subscriptions. The current level of support seems fine - with the caveat that _each subscription ideally has only one offer_.

**Comment by keesvandieren** [comment]
We have  the implementation now which alway selects the first.

[Add_support_for_subscriptions_with_free_trial.patch.zip](https://github.com/libgdx/gdx-pay/files/9661664/Add_support_for_subscriptions_with_free_trial.patch.zip)

Ok with this PR?

I think I cannot push to `mathkend:master`? Can @mathkend apply this patch, update the PR? Or approve the patch and I will create a new PR with all the changes in it.

**Comment by kharindev** [comment]
Do you think it will be possible to update Google Billing by November?

**Comment by keesvandieren** [comment]
@kharindev  yes, if I get no feedback on my latest comment I will be creating a new PR with all fixes soon and release it.

**Comment by keesvandieren** [comment]
Completed with PR #249

**Pull request closed without merging**